flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/FlinkCalciteCatalogReader.java

```diff
@@ -62,17 +80,181 @@ public FlinkCalciteCatalogReader(
 		if (originRelOptTable == null) {
 			return null;
 		} else {
-			// Wrap FlinkTable as FlinkRelOptTable to use in query optimization.
-			FlinkTable table = originRelOptTable.unwrap(FlinkTable.class);
+			// Wrap as FlinkPreparingTableBase to use in query optimization.
+			CatalogSchemaTable table = originRelOptTable.unwrap(CatalogSchemaTable.class);
 			if (table != null) {
-				return FlinkRelOptTable.create(
-					originRelOptTable.getRelOptSchema(),
-					originRelOptTable.getRowType(),
+				return toPreparingTable(originRelOptTable.getRelOptSchema(),
 					originRelOptTable.getQualifiedName(),
+					originRelOptTable.getRowType(),
 					table);
 			} else {
 				return originRelOptTable;
 			}
 		}
 	}
+
+	/**
+	 * Translate this {@link CatalogSchemaTable} into Flink source table.
+	 */
+	private static FlinkPreparingTableBase toPreparingTable(
+			RelOptSchema relOptSchema,
+			List<String> names,
+			RelDataType rowType,
+			CatalogSchemaTable table) {
+		if (table.isTemporary()) {
```

Comment:
If i'm not mistaken, `convertTemporaryTable` and `convertPermanentTable` have exactly the same logic here.

Reply:
I agree with Kurt. The logic is the same now. In the original code, the difference between them is the stats will be UNKNOW if it is temporary. However, we moved the stats logic into `DatabaseCalciteSchema#getTable`, so I think we don't need to distinguish `convertTemporaryTable` and `convertPermanentTable` here.

Reply:
Yep, thanks, let me refactor one of them out !!!